### PR TITLE
test(eval): expand basic suite with 4 practical agent capability tests

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary
- Add 4 new eval tests to the basic suite, expanding from 6 to 10 tests
- Tests cover real-world agent workflows not previously evaluated:
  - **json-transform**: Parse JSON data, compute derived values, output structured result
  - **multi-file-refactor**: Rename a function across multiple files while preserving behavior
  - **write-tests**: Given library code, write comprehensive pytest tests with edge cases
  - **fix-import-error**: Trace dependency chain and create missing module to fix imports

## Motivation
Addresses gaps identified in #63 — the existing eval suite focuses on basic file creation, patching, and single-file operations. These new tests exercise:
- **Data processing** (JSON I/O with computation)
- **Multi-file operations** (coordinated changes across files)
- **Test writing** (a common agent task — given code, produce tests)
- **Dependency debugging** (understanding import chains, creating missing pieces)

All check functions are validated with expected inputs. No new dependencies added.

## Test plan
- [x] All 10 tests load successfully (`from gptme.eval.suites.basic import tests`)
- [x] All check functions verified with expected input/output pairs
- [x] Ruff check passes
- [x] mypy passes

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by ErikBjare** [comment]
@TimeToBuildBob Address review

File: gptme/eval/suites/basic.py [hunk 1]
```diff
@@ -62,6 +62,85 @@ def check_read_modify_file(ctx):
...
+    """utils.py should define calculate_total instead of calcTotal."""
+    content = ctx.files.get("utils.py", "")
+    if isinstance(content, bytes):
+        content = content.decode()
+    return "def calculate_total" in content and "def calcTotal" not in content
+
+
+def check_refactor_output(ctx):
+    """Program should still produce correct output after refactoring."""
+    return "150" in ctx.stdout.split()
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in ea2e1452b — changed to substring matching (`"150" in ctx.stdout`), consistent with `check_read_modify_output`.

**Comment by TimeToBuildBob** [comment]
## Greptile Review Addressed

Fixed the critical bug identified by Greptile in `check_refactor_output` (commit ea2e1452b):

**Bug**: `"150" in ctx.stdout.split()` — list membership check fails because Python outputs `"150.0"` (float), not `"150"`.

**Fix**: Changed to `"150" in ctx.stdout` — substring matching, consistent with `check_read_modify_output`.

CI is passing (lint, typecheck, build, Docker all green; API key tests still pending as expected).

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/eval/suites/basic.py [hunk 2]
```diff
@@ -62,6 +62,85 @@ def check_read_modify_file(ctx):
...
+def check_json_transform_output(ctx):
+    """Output should contain the transformed JSON data with total revenue."""
+    import json
+
+    try:
+        data = json.loads(ctx.stdout.strip())
+    except (json.JSONDecodeError, ValueError):
+        return False
+    # products: A(10*100=1000), B(20*50=1000), C(5*200=1000) => total 3000
+    return data.get("total_revenue") == 3000
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in cb3a672 — `check_json_transform_output` now validates both `total_revenue == 3000` and the `products` list (correct length and per-product revenues A=1000, B=1000, C=1000). A partial output missing the products list will now fail.

File: gptme/eval/suites/basic.py [hunk 3]
```diff
@@ -62,6 +62,85 @@ def check_read_modify_file(ctx):
...
+    """pytest should report multiple tests passed."""
+    return "passed" in ctx.stdout.lower() or "passed" in ctx.stderr.lower()
+
+
+# --- fix-import-error checks ---
+
+
+def check_fix_import_output(ctx):
+    """After fixing, the program should output the correct result."""
+    return "42" in ctx.stdout.split()
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in cb3a672 — `check_fix_import_output` was missing its function definition entirely (syntax error), which is what caused the lint failure. It already used `"42" in ctx.stdout` (substring matching), not `.split()`. The function definition has been restored.

**Comment by TimeToBuildBob** [comment]
## All Greptile Review Comments Addressed (cb3a672)

- **`check_fix_import_output` syntax error**: Function definition was accidentally dropped, leaving a bare `return` statement outside any function — this was the lint failure. Restored in cb3a672.
- **`check_json_transform_output` incomplete validation**: Now validates both `total_revenue == 3000` and the `products` list (3 items with correct per-product revenues). Partial output without the products breakdown will correctly fail.
- **List-membership pattern**: `check_fix_import_output` was already using substring matching (`"42" in ctx.stdout`), not `.split()` — no change needed there.

Triggering a fresh Greptile review.

@greptileai review

File: gptme/eval/suites/basic.py [hunk 4]
```diff
@@ -62,6 +62,92 @@ def check_read_modify_file(ctx):
...
+
+# --- multi-file-refactor checks ---
+
+
+def check_refactor_main(ctx):
+    """main.py should call calculate_total instead of calcTotal."""
+    content = ctx.files.get("main.py", "")
+    if isinstance(content, bytes):
+        content = content.decode()
+    return "calculate_total" in content and "calcTotal" not in content
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in 878d186f6 — changed to `"calcTotal(" not in content` to anchor the check to function calls/definitions only, preventing false-negatives from agent comments like `# was calcTotal`. Now consistent with `check_refactor_utils` which anchors to `"def calcTotal"`.

File: gptme/eval/suites/basic.py [hunk 5]
```diff
@@ -62,6 +62,92 @@ def check_read_modify_file(ctx):
...
+    return "test_mathlib.py" in ctx.files
+
+
+def check_tests_pass(ctx):
+    return ctx.exit_code == 0
+
+
+def check_tests_output(ctx):
+    """pytest should report multiple tests passed."""
+    return "passed" in ctx.stdout.lower() or "passed" in ctx.stderr.lower()
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in 878d186f6 — replaced the redundant `"passed" in stdout` check with a function-coverage check: `all(fn in ctx.stdout for fn in ("factorial", "is_palindrome", "clamp"))`. Since the run command is `pytest -v`, all three function names will appear in the verbose output only if the agent wrote tests for each of them. This is now a meaningful independent assertion that `check_tests_pass` doesn't subsume.

**Comment by TimeToBuildBob** [comment]
## Greptile Review Addressed (878d186f6)

- **`check_refactor_main` false-negative**: Changed `"calcTotal" not in content` → `"calcTotal(" not in content` to anchor the check to function calls/definitions. An agent comment like `# was calcTotal` won't cause a false-negative. Consistent with `check_refactor_utils` which already anchors to `"def calcTotal"`.

- **`check_tests_output` redundancy**: Replaced `"passed" in stdout` (always true when `check_tests_pass` passes) with `all(fn in ctx.stdout for fn in ("factorial", "is_palindrome", "clamp"))`. With `-v`, pytest prints each collected test name, so all three function names appear in output only if the agent wrote tests covering all three library functions — a meaningful independent assertion.

@greptileai review

File: gptme/eval/suites/basic.py [hunk 6]
```diff
@@ -62,6 +62,92 @@ def check_read_modify_file(ctx):
...
+    """All three library functions should have tests (visible in verbose pytest -v output)."""
+    return all(fn in ctx.stdout for fn in ("factorial", "is_palindrome", "clamp"))
+
+
+# --- fix-import-error checks ---
+
+
+def check_fix_import_output(ctx):
+    """multiply(6, 7) should output 42; use substring match to handle 42.0 float output."""
+    return "42" in ctx.stdout
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in 20be1c781 — changed to `ctx.stdout.strip() in ("42", "42.0")` for exact match. Eliminates the false-positive on outputs like "420" while still handling float output.

File: gptme/eval/suites/basic.py [hunk 7]
```diff
@@ -62,6 +62,92 @@ def check_read_modify_file(ctx):
...
+    return "test_mathlib.py" in ctx.files
+
+
+def check_tests_pass(ctx):
+    return ctx.exit_code == 0
+
+
+def check_tests_output(ctx):
+    """All three library functions should have tests (visible in verbose pytest -v output)."""
+    return all(fn in ctx.stdout for fn in ("factorial", "is_palindrome", "clamp"))
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in 20be1c781 — switched from checking pytest stdout to inspecting `ctx.files["test_mathlib.py"]` directly. All three identifiers (`factorial`, `is_palindrome`, `clamp`) must appear in the test file source, which is true regardless of how the agent structures its test functions (grouped, parametrized, class-based, etc.).

**Comment by TimeToBuildBob** [comment]
## Final Greptile Review Addressed (20be1c781)

- **`check_fix_import_output` false-positive on \"420\"**: Changed `\"42\" in ctx.stdout` → `ctx.stdout.strip() in (\"42\", \"42.0\")`. Exact match — a buggy `multiply` returning 420 no longer silently passes.

- **`check_tests_output` false-negative for grouped tests**: Changed from checking pytest `-v` stdout (which requires function names in pytest output) → inspecting `ctx.files[\"test_mathlib.py\"]` directly. All three identifiers (`factorial`, `is_palindrome`, `clamp`) must appear in the test source, independent of test organization style.

@greptileai review

File: gptme/eval/suites/basic.py [hunk 8]
```diff
@@ -62,6 +62,100 @@ def check_read_modify_file(ctx):
...
+
+# --- multi-file-refactor checks ---
+
+
+def check_refactor_main(ctx):
+    """main.py should call calculate_total instead of calcTotal."""
+    content = ctx.files.get("main.py", "")
+    if isinstance(content, bytes):
+        content = content.decode()
+    return "calculate_total" in content and "calcTotal(" not in content
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py]: Fixed in 8ec8ca25a — added `"from utils import calculate_total" in content` to the check. An agent that updates only the call site but leaves `from utils import calcTotal` in place will now correctly fail.

File: gptme/eval/suites/basic.py [hunk 9]
```diff
@@ -161,4 +255,123 @@ def check_read_modify_file(ctx):
...
+            "Running 'python app.py' fails with an import error because math_ops.py doesn't exist. "
+            "Read the existing files to understand what's needed, "
+            "then create the missing math_ops.py module with the required function "
+            "so that app.py runs successfully and prints the correct result."
+        ),
+        "tools": ["read", "save", "shell"],
+        "expect": {
+            "correct output": check_fix_import_output,
+            "clean exit": check_fix_import_exit,
+        },
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py:383]: Fixed in 8ec8ca25a — added `check_fix_import_created` which verifies `"math_ops.py" in ctx.files`. An agent that inlines the multiply logic into helpers.py and skips creating the module will now fail this check.

**Comment by TimeToBuildBob** [comment]
## Greptile Review Addressed (8ec8ca25a)

- **`check_refactor_main` missing import check**: Added `"from utils import calculate_total" in content` to the check. Previously only anchored to call sites (`calcTotal(` not in content), missing the import line. An agent that updates the call but not the import now correctly fails.

- **`fix-import-error` missing module-creation check**: Added `check_fix_import_created` which asserts `"math_ops.py" in ctx.files`. The task explicitly requires creating the module; an agent that inlines the logic into helpers.py instead will now fail this check.

@greptileai review

File: gptme/eval/suites/basic.py [hunk 10]
```diff
@@ -62,6 +62,109 @@ def check_read_modify_file(ctx):
...
+def check_refactor_main(ctx):
+    """main.py should import and call calculate_total instead of calcTotal."""
+    content = ctx.files.get("main.py", "")
+    if isinstance(content, bytes):
+        content = content.decode()
+    return (
+        "from utils import calculate_total" in content
+        and "calculate_total" in content
+        and "calcTotal(" not in content
+    )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/eval/suites/basic.py:105]: Fixed in e87f6df — removed the redundant `and "calculate_total" in content` condition, which was fully subsumed by the import-line check above it.

**Comment by TimeToBuildBob** [comment]
## All Greptile Feedback Addressed (e87f6df)

- **Redundant condition in `check_refactor_main`**: Removed `and "calculate_total" in content` — fully subsumed by the `"from utils import calculate_total" in content` check above it.

Greptile's last review (4/5) flagged this as the only remaining issue and called the PR safe to merge.
